File: cloudflare-api/src/match-clock.ts
import type { MatchPhase, MatchRow, MatchStatus } from "./types";

export type MatchPhaseAction = "start_match" | "halftime" | "start_second_half" | "start_extra_time" | "finish_match";

export type MatchClockTransition = Pick<MatchRow, "status" | "phase" | "phase_started_at">;
type MatchClockInput = Pick<MatchRow, "status" | "phase" | "phase_started_at"> & Partial<Pick<MatchRow, "has_extra_time">>;

export class MatchPhaseTransitionError extends Error {
  readonly code: "invalid_match_phase" | "invalid_match_status";

  constructor(code: "invalid_match_phase" | "invalid_match_status") {
    super(code === "invalid_match_phase"
      ? "That match phase change is not available from the current state."
      : "Use the live scoring controls to change the match status in order.");
    this.code = code;
    this.name = "MatchPhaseTransitionError";
  }
}

export function transitionMatchPhase(
  match: MatchClockInput,
  action: MatchPhaseAction,
  occurredAt: string,
): MatchClockTransition {
  const allowed: Record<MatchPhaseAction, { status: MatchStatus; phases: MatchPhase[] }> = {
    start_match: { status: "scheduled", phases: ["not_started"] },
    halftime: { status: "live", phases: ["first_half"] },
    start_second_half: { status: "live", phases: ["halftime"] },
    start_extra_time: { status: "live", phases: ["second_half"] },
    finish_match: { status: "live", phases: ["first_half", "halftime", "second_half", "extra_time"] },
  };
  const expected = allowed[action];
  if (match.status !== expected.status || !expected.phases.includes(match.phase) || (action === "start_extra_time" && match.has_extra_time === 0)) {
    throw new MatchPhaseTransitionError("invalid_match_phase");
  }

  if (action === "start_match") return { status: "live", phase: "first_half", phase_started_at: occurredAt };
  if (action === "halftime") return { status: "live", phase: "halftime", phase_started_at: null };
  if (action === "start_second_half") return { status: "live", phase: "second_half", phase_started_at: occurredAt };
  if (action === "start_extra_time") return { status: "live", phase: "extra_time", phase_started_at: occurredAt };
  return { status: "finished", phase: "finished", phase_started_at: null };
}

export function transitionLegacyStatus(
  match: MatchClockInput,
  nextStatus: MatchStatus,
  occurredAt: string,
): MatchClockTransition {
  if (nextStatus === match.status) return { status: match.status, phase: match.phase, phase_started_at: match.phase_started_at };
  if (match.status === "scheduled" && nextStatus === "live") return transitionMatchPhase(match, "start_match", occurredAt);
  if (match.status === "live" && nextStatus === "finished") return transitionMatchPhase(match, "finish_match", occurredAt);
  throw new MatchPhaseTransitionError("invalid_match_status");
}

File: backend/app/services/match_clock.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Literal

from app.core.errors import api_error
from app.db.models import Match, MatchPhase, MatchStatus

MatchPhaseAction = Literal[
    "start_match",
    "halftime",
    "start_second_half",
    "start_extra_time",
    "finish_match",
]


def apply_phase_action(
    match: Match,
    action: MatchPhaseAction,
    *,
    occurred_at: datetime | None = None,
) -> None:
    now = occurred_at or datetime.now(UTC)
    expected: dict[MatchPhaseAction, tuple[MatchStatus, set[MatchPhase]]] = {
        "start_match": (MatchStatus.scheduled, {MatchPhase.not_started}),
        "halftime": (MatchStatus.live, {MatchPhase.first_half}),
        "start_second_half": (MatchStatus.live, {MatchPhase.halftime}),
        "start_extra_time": (MatchStatus.live, {MatchPhase.second_half}),
        "finish_match": (
            MatchStatus.live,
            {
                MatchPhase.first_half,
                MatchPhase.halftime,
                MatchPhase.second_half,
                MatchPhase.extra_time,
            },
        ),
    }
    expected_status, expected_phases = expected[action]
    if (
        match.status != expected_status
        or match.phase not in expected_phases
        or (action == "start_extra_time" and not match.has_extra_time)
    ):
        raise api_error(
            409,
            "invalid_match_phase",
            "That match phase change is not available from the current state.",
        )

    if action == "start_match":
        match.status = MatchStatus.live
        match.phase = MatchPhase.first_half
        match.phase_started_at = now
    elif action == "halftime":
        match.phase = MatchPhase.halftime
        match.phase_started_at = None
    elif action == "start_second_half":
        match.phase = MatchPhase.second_half
        match.phase_started_at = now
    elif action == "start_extra_time":
        match.phase = MatchPhase.extra_time
        match.phase_started_at = now
    else:
        match.status = MatchStatus.finished
        match.phase = MatchPhase.finished
        match.phase_started_at = None


def apply_legacy_status_change(match: Match, next_status: MatchStatus) -> None:
    if next_status == match.status:
        return
    if match.status == MatchStatus.scheduled and next_status == MatchStatus.live:
        apply_phase_action(match, "start_match")
        return
    if match.status == MatchStatus.live and next_status == MatchStatus.finished:
        apply_phase_action(match, "finish_match")
        return
    raise api_error(
        409,
        "invalid_match_status",
        "Use the live scoring controls to change the match status in order.",
    )
